Port to babashka and other things

arichiardi opened the pull request:
Coming back from magnificent adventure I decided to take up the classic code-on-airplane project and port this project to babashka.

Please review commit by commit as the commit message describes the changes.

arichiardi commented on src/envoy/tools.cljc:
```diff
@@ -47,22 +47,23 @@
   "consul values are strings. str->value will convert:
   the numbers to longs, the alphanumeric values to strings, and will use Clojure reader for the rest
   in case reader can't read OR it reads a symbol, the value will be returned as is (a string)"
+  [v & [deserializer]]
```
this was a clj-kondo warning (docstring goes before params)

arichiardi commented on src/envoy/core.cljc:
```diff
@@ -0,0 +1,229 @@
+                                {})))))
+                 {})
+         ;; make sure we get rid of empty keys (or the diff will be wrong)
+         (tools/remove-empty-keys))))
```
Nota Bene

arichiardi committed:
Quality of life refactor and warning removal

This patch removes a couple of warnings for unused variables and
refactors with-ops, recurse and friends to envoy.tools to avoid
duplication.
It also refactors the Base64 decoder to avoid recreating it at ever
fromBase64 invocation.
Finally, it removes unused options (like keywordize in format-resp).

arichiardi committed:
Normalize thrown exceptions

This patch consolidates exceptions around these things:
* Wrap actual exceptions with ex-info where the original is set as
  ex-cause.
* Never include maps in exception messages but use ex-data instead.

arichiardi committed:
Port to babashka-compatible code

Coming back from magnificent adventure I decided to take up the classic
code-on-airplane project and port this project to babashka.  The only
significant change has been to tweak the request timeout check and avoid
using the org.httpkit.client.TimeoutException class directly cause
babashka does not seem to include it.

tolitius commented on src/envoy/core.cljc:
```diff
@@ -0,0 +1,228 @@
+    (if-not overwrite?
+      (doseq [[k v] (tools/map->props m serializer)]
+        (put (str path "/" k) (str v) (dissoc ops :serializer :update)))
+      (overwrite-with2 path m ops))))
```
this looks identical to the `map->consul`
besides the `overwrite-with2` call

see the `overwrite-with2` comment :)

arichiardi replied:
Ok added a commit that fixes existing bugs and removes this function

tolitius commented on src/envoy/core.cljc:
```diff
@@ -0,0 +1,228 @@
+         ;; make sure we get rid of empty keys (or the diff will be wrong)
+         (tools/remove-empty-keys))))
+
+(defn- overwrite-with2
+  [path input-map & [{:keys [offset serializer] :or {serializer :edn} :as ops}]]
+  (let [current-map (compute-current-map path input-map ops)
+        ;; to update correctly seq we need to pre-serialize map
+        [to-add to-remove _] (diff (tools/serialize-map input-map serializer)
+                                   (tools/serialize-map current-map serializer))
+        side-effect-ops (dissoc ops :serializer :update)]
+    ;;add
+    (doseq [[k v] (tools/map->props to-add serializer)]
+      (put (tools/complete-key-path path offset k)
+           (str v)
+           side-effect-ops))
+    ;;remove
+    (doseq [[k _] (tools/map->props to-remove serializer)]
+      (delete (tools/complete-key-path path offset k)
+              side-effect-ops))))
```
if this does what `overwrite-with` already does (+ the `offset`), than I don't think we need to create `overwrite-with2`, but rather keep it as `overwrite-with`, given that it does 100% of what the current `overwrite-with` does.

I understand `foo`, `foo2` talk by Rich, but names are too important to sacrifice
if we have a library with:

foo2
foo
bar3
bar
foo9
bar4

it would be quite difficult to select a proper "`foo`" and "`bar`" for the use case
even more difficult to make sense of the dependencies between foo13 and bar7

the reason is "`2`" or "`13`" convey little business context unless it is 2 miles per hour or $13.

arichiardi replied:
Yeah I do agree with you. 

The reason for that `2` is that I could not figure out a problem with the current version and I went for the lazy way of writing my own implementation. I'll try again and see - I was getting a "path not found". I will get back to you.

arichiardi replied:
I think I found the bug - that `delete` in `overwrite-with` can never work because it does not carry the auth token - will push a small separate commit and move things around.

arichiardi replied:
Going back to the naming thing probably a namespace2 makes more sense generally for achieving accretion

arichiardi replied:
Ok added a commit that fixes existing bugs and removes this function

tolitius commented on src/envoy/session.cljc:
```diff
@@ -1,59 +1,56 @@
+   (let [{:keys [body error status] :as resp} @(http/get path
+                                                         (tools/with-ops ops))]
+     (if (or error (not= status 200))
+       (throw (ex-info (str "could not GET from consul due to: " (error-message resp))
```
👍 
thanks for these

arichiardi committed:
Fix overwrite-with implementation

A couple of problems were detected and fixed:

1) The authentication token was not passed down to delete and that
   resulted in 404s.

2) Deleting was down after adding and that would not work with the way
   clojure.data/diff identifies differences (based on value, see inline
   comment).

The new implementation now correctly adds and deletes entries.

arichiardi commented on src/envoy/core.clj:
```diff
@@ -101,32 +101,47 @@
+           side-effect-ops))))
 
 (defn map->consul
-  [kv-path m & [{:keys [serializer overwrite?] :or {serializer :edn overwrite? false} :as ops}]]
```
This is just a renaming of `kv-path` -> `path` and `m` -> `input-map`

arichiardi committed:
Avoid using clojure's read-string

It is known to be a security risk as it can execute code rather than
just un-marshaling its edn input. This patch swaps it with
`clojure.edn/read-string` which is safer.

Also see warning at this page:

  https://clojuredocs.org/clojure.edn

tolitius commented:
thank you!

tolitius commented:
changes should be in:
```
{tolitius/envoy {:mvn/version "0.1.33-SNAPSHOT"}}
```
